File: utils/src/main/java/ru/virgil/utils/fluent_request/Requester.java
package ru.virgil.utils.fluent_request;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.RequiredArgsConstructor;
import org.springframework.http.MediaType;
import org.springframework.stereotype.Component;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.MvcResult;
import org.springframework.test.web.servlet.request.MockHttpServletRequestBuilder;
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders;
import ru.virgil.utils.TestUtils;

import java.util.Optional;

@RequiredArgsConstructor
@Component
class Requester {

    private final MockMvc mockMvc;
    private final TestUtils testUtils;
    private final ObjectMapper jackson = new ObjectMapper();

    private MockHttpServletRequestBuilder getHttpBuilder(RequestMethod requestMethod, String url) {
        return switch (requestMethod) {
            case GET -> MockMvcRequestBuilders.get(url);
            case POST -> MockMvcRequestBuilders.post(url);
            case PUT -> MockMvcRequestBuilders.put(url);
            case DELETE -> MockMvcRequestBuilders.delete(url);
        };
    }

    private void addBody(MockHttpServletRequestBuilder builder, Object objectBody) throws JsonProcessingException {
        String jsonBody = jackson.writeValueAsString(objectBody);
        builder.content(jsonBody)
                .contentType(MediaType.APPLICATION_JSON);
    }

    protected MvcResult makeRequest(RequestModel requestModel) throws Exception {
        MockHttpServletRequestBuilder builder = getHttpBuilder(requestModel.getRequestMethod(), requestModel.getUrl());
        if (Optional.ofNullable(requestModel.getRequestBody()).isPresent()) {
            addBody(builder, requestModel.getRequestBody());
        }
        return mockMvc.perform(builder)
                .andExpectAll(requestModel.getResultMatchers())
                .andDo(testUtils::printResponse)
                .andReturn();
    }
}

File: tools/src/main/kotlin/ru/virgil/spring_tools/tools/testing/fluent/request/Requester.kt
package ru.virgil.spring_tools.tools.testing.fluent.request

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.kotlin.readValue
import org.springframework.http.HttpMethod
import org.springframework.http.MediaType
import org.springframework.mock.web.MockMultipartFile
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.MvcResult
import org.springframework.test.web.servlet.ResultMatcher
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.status
import org.springframework.util.MimeType
import org.springframework.util.MimeTypeUtils
import ru.virgil.spring_tools.tools.testing.TestUtils
import java.net.URI

class Requester(
    val objectMapper: ObjectMapper,
    val mockMvc: MockMvc,
    val testUtils: TestUtils,
    var config: RequestConfig = RequestConfig(),
) {

    fun get(function: Requester.() -> String) = setConnectionOptions(function, HttpMethod.GET)

    fun post(function: Requester.() -> String) = setConnectionOptions(function, HttpMethod.POST)

    fun put(function: Requester.() -> String) = setConnectionOptions(function, HttpMethod.PUT)

    fun delete(function: Requester.() -> String) = setConnectionOptions(function, HttpMethod.DELETE)

    private fun setConnectionOptions(uriFunction: Requester.() -> String, httpMethod: HttpMethod) {
        config = config.copy(httpMethod = httpMethod, uri = URI.create(uriFunction(this)))
    }

    fun send(function: Requester.() -> Any) {
        config = config.copy(dto = function(this))
    }

    fun expect(function: Requester.() -> ResultMatcher) {
        config.expects.add(function(this))
    }

    fun file(function: Requester.() -> MockMultipartFile) {
        config = config.copy(file = function(this))
    }

    inline fun <reified T> makeRequest(): T {
        val config = config.copy()
        val requestBuilder = when {
            config.httpMethod == null || config.uri == null ->
                throw Exception("Нужно указать ${HttpMethod::class.simpleName} и ${URI::class.simpleName}")
            config.file != null && config.dto != null ->
                throw Exception("Нужно отправлять что-то одно: DTO или ${MockMultipartFile::class.simpleName}")
            config.file != null -> MockMvcRequestBuilders.multipart(config.httpMethod, config.uri)
                .file(config.file)
            config.dto != null -> MockMvcRequestBuilders.request(config.httpMethod, config.uri)
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(config.dto))
            else -> MockMvcRequestBuilders.request(config.httpMethod, config.uri)
        }
        if (config.expects.isEmpty()) {
            config.expects.add(status().isOk)
        }
        val result = mockMvc.perform(requestBuilder)
            .andExpectAll(*config.expects.toTypedArray())
            .andDo { mvcResult: MvcResult -> testUtils.printResponse(mvcResult) }
            .andReturn()
        val mimeType = MimeTypeUtils.parseMimeType(result.response.contentType ?: MimeTypeUtils.TEXT_PLAIN_VALUE)
        return when {
            isJson(mimeType, result) -> objectMapper.readValue(result.response.contentAsString)
            isImage(mimeType) -> result.response.contentAsByteArray as T
            else -> result as T
        }
    }

    fun isImage(mimeType: MimeType) =
        mimeType.isPresentIn(listOf(MimeTypeUtils.IMAGE_GIF, MimeTypeUtils.IMAGE_JPEG, MimeTypeUtils.IMAGE_PNG))

    fun isJson(mimeType: MimeType, result: MvcResult) =
        mimeType.isPresentIn(listOf(MimeTypeUtils.APPLICATION_JSON)) && result.response.contentAsString.isNotBlank()

}
